### ISSUE TITLE
"Permission Denied"

### ISSUE THREAD

**Issue by Banane6077** [ansible-semaphore/semaphore#1518]
Hello fellow People...

i am kinda new to ansible/semaphore and tryin' to set up my first playbook.

My Playbook:

"---
- hosts: all
  become: yes
  become_user: myuser
  tasks:
    - name: Update apt repo and cache on all Debian/Ubuntu boxes
      apt: update_cache=yes force_apt_get=yes cache_valid_time=3600

    - name: Upgrade all packages on servers
      apt: upgrade=dist force_apt_get=yes

    - name: Check if a reboot is needed on all servers
      register: reboot_required_file
      stat: path=/var/run/reboot-required get_md5=no

    - name: Reboot the box if kernel updated
      reboot:
        msg: "Reboot initiated by Ansible for kernel updates"
        connect_timeout: 5
        reboot_timeout: 300
        pre_reboot_delay: 0
        post_reboot_delay: 30
        test_command: uptime
      when: reboot_required_file.stat.exists
"

My error log is in the attachement...

i made already an "ansible all -m ping" and i get a success
even the connection via "ssh user@hostip" works fine and without needing a password.

Can you guys help me with that?

[Ansible Error.txt](https://github.com/ansible-semaphore/semaphore/files/12726942/Ansible.Error.txt)

**Comment by hmoffatt** [comment]
You've set `become_user: myuser`, but only root can run the `apt` and `reboot` tasks. The `apt` fails first because you're running it as non-root. (This isn't anthing to do with Semaphore.)

**Comment by Banane6077** [comment]
> You've set `become_user: myuser`, but only root can run the `apt` and `reboot` tasks. The `apt` fails first because you're running it as non-root. (This isn't anthing to do with Semaphore.)

Hello! Thank you for your input!
I still have the same errors...


8:32:10 AM
TASK [Update apt repo and cache on all Debian/Ubuntu boxes] ********************
8:33:11 AM
fatal: [172.17.3.15]: FAILED! => {"changed": false, "msg": "Failed to lock apt for exclusive operation: Failed to lock directory /var/lib/apt/lists/: E:Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)"}

**Comment by hmoffatt** [comment]
You need `become: yes` with `become_user: root` (or omitted).

**Comment by fuuman** [comment]
But if I have a SSH-Key copied for user A and user A is allowed to execute "sudo" commands and everything is working manually, why isn't it possible to execute everything as non-root user? I am having the same error.

**Comment by hmoffatt** [comment]
Post your whole playbook.
Also, does the playbook run on the command line without semaphore?

**Comment by fuuman** [comment]
Playbook: https://pastebin.com/An6wM4rZ

Yes, on the commandline it runs without any issues. Using -K and providing the password manually.

**Comment by fiftin** [comment]
Fixing in progress

**Comment by fuuman** [comment]
Is there an older version of the docker image without this bug? So that I could use this in the meantime?

**Comment by Kimbaras** [comment]
Hey all,

Is there an ETA on the fix? 

I believe I've stumbled on the same / similar issue as explained in the mentioned issue.
Basically it looks like the "become" part is being ignored when executed by Semaphore while it's working when executing the playbook manually from cli (as reported also above by another user).

When I tried to "force" the sudo / become part (a simple _command: sudo touch file_) the task hangs indifinetely.

 @fuuman not sure if this is helpful to you, but on a different system (running Ansible 2.9.27 and Semaphore 2.8.90) this seems to be working fine

Thanks!

**Comment by Kimbaras** [comment]
Hi all,

I've done some testing about this issue and tried previous versions of Semaphore to see if I can get an older version to work while a fix is being worked on.

I've tried the following Semaphore versions

- 2.9.37
- 2.9.4
- 2.8.90

On all these versions the issue with the sudo / become seems present.
My system is an Ubuntu 22.04 running Ansible 2.15.4

The strange thing is that on a different system, version 2.8.90 is working correctly (Ansible 2.9.27 - Semaphore 2.8.90 running on Raspberry with Raspbian (Buster)).

Between the two systems the differences are:

- The Ansible version (2.15.4 - not working, 2.9.27 - working)
- The fact that on the Ubuntu 22.04 system I'm using LDAP accounts.

Please note that running the playbook manually from cli (ansible-playbook...) is working as expected.
So it seems it's not an issue of Ansible, but the interaction between Semaphore <-> Ansible (2.15.4).

Unfortunately at the moment I'm unable to rollback the Ansible version to test with a previous version.

Hoping this helps and a fix / workaround arrives fast!

Thanks!

**Comment by innovortex** [comment]
I have the same issue, even with the latest beta.
Is a fix in sight?

**Comment by twistedlogick** [comment]
I don't know if this will help anyone else, but this seemed to fix my issues with not being able to run commands as sudo - remove the username from the login with password key in the keystore I was using for sudo-ing:
https://github.com/ansible-semaphore/semaphore/issues/1445#issuecomment-1736925703

**Comment by bilalbayasut** [comment]
I got the same problem. 

```
...
ERROR: Unable to lock database: Permission denied
ERROR: Failed to open apk database: Permission denied
could not update package db
```

In my setup, because semaphore (docker) uses alpine, I have to change `become_method = su` in the `ansible.cfg` and also got to change my task from using apt into using apk in the playbook.

I have tried workaround from @twistedlogick but it shows another error complaining no username found :) . I hope this is because this bug

**Comment by wuyue92tree** [comment]
Maybe work

My playbook use to clean docker unused images & container which also need root permission.

semaphore version is v2.9.37

```
- name: clean docker
  hosts: docker
  
  tasks:
    - name: docker clean
      become: true
      become_user: root
      become_method: sudo
      command: docker system prune -a -f
```

at first it's always failed,  finaly I found an wrong config in my `Inventory`, I set both `User Credentials` and `Sudo Credentials (Optional)` to the same `Key Store`, this is the fail reason.  the `Sudo Credentials (Optional)` should not have a username I guest. When I create a new `Key Store` keep the username none, it' worked.

wrong
![image](https://github.com/ansible-semaphore/semaphore/assets/7899044/42493319-b080-48ef-8d93-71c5c7ebbbde)

right
![image](https://github.com/ansible-semaphore/semaphore/assets/7899044/8f755b5f-3285-4fa9-b1dc-070c24c58b90)


Good luck~

**Comment by bilalbayasut** [comment]
strangely enough that works, thanks @wuyue92tree  . so to wrap up. two keystores need to be created
1. one keystore entry (login_password or ssh) that contains username and password (though the password wont be used in this keystore)
2. one keystore entry that only containing password, (emptying out the username).

and put those in the inventory, and use it as vault password in the template

**Comment by RusticPotatoes** [comment]
> strangely enough that works, thanks @wuyue92tree . so to wrap up. two keystores need to be created
> 
>     1. one keystore entry (login_password or ssh) that contains username and password (though the password wont be used in this keystore)
> 
>     2. one keystore entry that only containing password, (emptying out the username).
> 
> 
> and put those in the inventory, and use it as vault password in the template

can confirm works.. cheers

**Comment by PauliusGitSoft** [comment]
It works, googled two days. Thanks @wuyue92tree

**Comment by cbrookins** [comment]
Thanks for sharing.  This helped me out.

**Comment by ghost** [comment]
This fixed it for me, too.  Thank you

**Comment by dcolley** [comment]
@wuyue92tree 's reply works because of the sudo command

**UserId:**
You need to leave userid blank or put `root` (i.e. the user you want to become)

**Password:**
You need to put your password that allows you to call sudo

I think an update to the documentatio would allow this issue to be closed.

**Comment by bilalbayasut** [comment]
> strangely enough that works, thanks @wuyue92tree . so to wrap up. two keystores need to be created
> 
> 1. one keystore entry (login_password or ssh) that contains username and password (though the password wont be used in this keystore)
> 2. one keystore entry that only containing password, (emptying out the username).
> 
> and put those in the inventory, and use it as vault password in the template

this doesn't work anymore, it ended up working by putting the correct value in vault password (a entry with username and password filled) . did they fix this up ?

**Comment by DevOpsSim** [comment]
@fiftin Добрый день! Скажите, будет ли исправление проблемы повышения привилегий sudo или не ждать и не мучаться в итоге?

**Comment by fiftin** [comment]
@DevOpsSim в процессе исправления. Буду благодарен, если поможете в тестировании предварительной версии.

**Comment by fiftin** [comment]
@DevOpsSim https://github.com/semaphoreui/semaphore/releases/tag/v2.9.66-beta

```
docker pull semaphoreui/semaphore:v2.9.66-beta
```

UPDATE: this version has a bug. Will fix soon.

UPDATE: https://github.com/semaphoreui/semaphore/releases/tag/v2.9.68-beta

**Comment by Koltsz** [comment]
@wuyue92tree Thankyou!

This waas driving me mad

**Comment by fiftin** [comment]
Hi @Koltsz 
Do you have a use case to reproduce the problem? Step-by-step.

**Comment by Koltsz** [comment]
> Hi @Koltsz 
> Do you have a use case to reproduce the problem? Step-by-step.

Hey, sure

- Installed the latest docker image. Using mysql as the DB.
- Started with adding a key with type login with password called docker-server
- created Environment var and added "ANSIBLE_HOST_KEY_CHECKING": "False" so I could use login with password.
- Inventory is a static file with just one IP to test this is where the issue was but I'll explain what I had before I did the fix.
The user credentials are set to docker-server and the Sudo credentials were also set to docker-server.

- task set up with my GitHub, the playbook is called update-servers.yaml

Here is the link:
https://github.com/Koltsz/ansible/blob/main/playbooks%2Fupdate-servers.yaml


This is the error I got:
```
2:43:38 PM
Task 19 added to queue
2:43:43 PM
Started: 19
2:43:43 PM
Run TaskRunner with template: update-servers
2:43:43 PM
Preparing: 19
2:43:43 PM
Updating Repository https://github.com/Koltsz/ansible
2:43:43 PM
From https://github.com/Koltsz/ansible
2:43:43 PM
 * branch            main       -> FETCH_HEAD
2:43:43 PM
Already up to date.
2:43:43 PM
installing static inventory
2:43:43 PM
No collections/requirements.yml file found. Skip galaxy install process.
2:43:43 PM
No roles/requirements.yml file found. Skip galaxy install process.
2:43:44 PM
PLAY [all] *********************************************************************
2:43:44 PM
TASK [Gathering Facts] *********************************************************
2:43:44 PM
2:43:44 PM
2:43:45 PM
ok: [10.0.0.93]
2:43:45 PM
2:43:45 PM
TASK [upgrade apt packages] ****************************************************
2:44:45 PM
fatal: [10.0.0.93]: FAILED! => {"changed": false, "msg": "Failed to lock apt for exclusive operation: Failed to lock directory /var/lib/apt/lists/: E:Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)"}
2:44:45 PM
2:44:45 PM
PLAY RECAP *********************************************************************
2:44:45 PM
10.0.0.93                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
2:44:45 PM
2:44:45 PM
Running playbook failed: exit status 2
```

And I tried everything to fix it, the only way to fix it was for me to follow what @wuyue92tree explained to do.


That's pretty much it.

**Comment by fiftin** [comment]
@Koltsz 

Please try `docker pull semaphoreui/semaphore:v2.9.68-beta`.

I just did what you said and it is works.

FYI:

I changed method of passing login/password to `ansible-playbook`. Now Semaphore uses flags `--ask-pass`, `--ask-become-pass` and sends passwords via Stdin.

**Comment by DevOpsSim** [comment]
> [@DevOpsSim](https://github.com/DevOpsSim)в процессе исправления. Буду благодарен, если поможете в предварительной версии теста.



> [@DevOpsSim](https://github.com/DevOpsSim) в процессе исправления. Буду благодарен, если поможете в тестировании предварительной версии.

Проверю, отпишусь. Спасибо!

**Comment by DevOpsSim** [comment]
Обновил  на  [v2.9.68-beta](https://github.com/semaphoreui/semaphore/releases/tag/v2.9.68-beta)

Мой playbook:
---
- hosts: all
  gather_facts: no
  become: true
  become_user: ansible
  tasks:
    - name: make sure SSL certificates are existing and secured
      copy:
        src: "/tmp/semaphore/repository_1_1/deploy_ssl/v_1/ssl/{{ item }}/domain.crt"
        dest: "/home/ansible/ssl/{{ item }}/domain.crt"
        owner: ansible
        group: ansible
        mode: "0644"
      loop:
        - example_1.com
        - example_2.com
        - example_3.com

    - name: make sure SSL private keys are installed
      copy:
        src: "/tmp/semaphore/repository_1_1/deploy_ssl/v_1/ssl/{{ item }}/domain.key"
        dest: "/home/ansible/ssl/{{ item }}/domain.key"
        owner: ansible
        group: ansible
        mode: "0644"
      no_log: true
      loop:
        - example_1.com
        - example_2.com
        - example_3.com

    - name: Create symbolic link
      file:
        src: "/home/ansible/ssl"
        dest: "/etc/nginx/ssl"
        state: link
        
    - name: reload nginx
      service:
        name: nginx.service
        state: reloaded

Авторизация и выполнение под пользователем ansible. Ему даны права sudo. В инвентаре в semaphore  Учетные данные пользователя: логин и пароль пользователя ansible. Повышенные учетные данные(дополнительно): только пароль пользователя ansible, без логина.
P.S. если запускаю этот же playbook локально на машине, подключаюсь по ssh под этим же пользователем ansible. выполняется без ошибок. Ссылка создается, сервис nginx reloded. Все ОК! 

3:14:37 PM
PLAY [all] *********************************************************************
3:14:37 PM
3:14:37 PM
TASK [make sure SSL certificates are existing and secured] *********************
3:14:41 PM
changed: [test-debian] => (item=example_1.com)
3:14:46 PM
changed: [test-debian] => (item=example_2.com)
3:14:51 PM
changed: [test-debian] => (item=example_3.com)
3:14:51 PM
3:14:51 PM
TASK [make sure SSL private keys are installed] ********************************
3:14:55 PM
changed: [test-debian] => (item=None)
3:15:00 PM
changed: [test-debian] => (item=None)
3:15:04 PM
changed: [test-debian] => (item=None)
3:15:04 PM
changed: [test-debian]
3:15:04 PM
3:15:04 PM
TASK [Create symbolic link] ****************************************************
3:15:06 PM
fatal: [test-debian]: FAILED! => {"changed": false, "msg": "Error while linking: [Errno 13] Permission denied: b'/home/ansible/ssl' -> b'/etc/nginx/ssl'", "path": "/etc/nginx/ssl"}
3:15:06 PM
3:15:06 PM
PLAY RECAP *********************************************************************
3:15:06 PM
test-debian                : ok=3    changed=3    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
3:15:06 PM
3:15:06 PM
Running playbook failed: exit status 2

**Comment by fiftin** [comment]
Что если убрать `become_user: ansible`? Возможно тогда заработает.

**Comment by fiftin** [comment]
Сейчас логика работы Semaphore очень простая: если есть параметр Become User, то в `ansible-playbook` передаются параметры `--ask-become-pass`, если указан логин, то еще `--become-user`.

**Comment by DevOpsSim** [comment]
Убрал

---
- hosts: all
  gather_facts: no
  become_method: sudo
  become: true
  
  tasks:
    - name: make sure SSL certificates are existing and secured
      copy:
        src: "/tmp/semaphore/repository_1_1/deploy_ssl/v_1/ssl/{{ item }}/domain.crt"
        dest: "/home/ansible/ssl/{{ item }}/domain.crt"
        owner: ansible
        group: ansible
        mode: "0644"
      loop:
        - example_1.com
        - example_2.com
        - example_3.com

    - name: make sure SSL private keys are installed
      copy:
        src: "/tmp/semaphore/repository_1_1/deploy_ssl/v_1/ssl/{{ item }}/domain.key"
        dest: "/home/ansible/ssl/{{ item }}/domain.key"
        owner: ansible
        group: ansible
        mode: "0644"
      no_log: true
      loop:
        - example_1.com
        - example_2.com
        - example_3.com

    - name: Create symbolic link
      file:
        src: "/home/ansible/ssl"
        dest: "/etc/nginx/ssl"
        state: link
        
    - name: reload nginx
      service:
        name: nginx.service
        state: reloaded

Получил это:

3:38:57 PM
PLAY [all] *********************************************************************
3:38:57 PM
3:38:57 PM
TASK [make sure SSL certificates are existing and secured] *********************
3:38:59 PM
fatal: [test-debian]: FAILED! => {"msg": "Failed to get information on remote file (/home/ansible/ssl/example_1.com/domain.crt): Shared connection to 192.168.62.57 closed.\r\n"}
3:38:59 PM
3:38:59 PM
PLAY RECAP *********************************************************************
3:38:59 PM
test-debian                : ok=1    changed=1    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
3:38:59 PM
3:38:59 PM
Running playbook failed: exit status 2

**Comment by fiftin** [comment]
Вероятно директории `/home/ansible/ssl/example_1.com/` не существует.

**Comment by fiftin** [comment]
@DevOpsSim вероятно эта ошибка: https://www.tecmint.com/fix-shared-connection-to-x-x-xx-closed-ansible-error/

Но это не относится к Semaphore, а относится к окружению (если вы запускаете Semaphore в Docker-е): версия python, ansible, etc. Вы можете поставить галочку `Debug` при запуске задачи и сравнить версии с локальным окружением и версии python на удаленном сервере.

**Comment by DevOpsSim** [comment]
@fiftin  Существует! Ладно... для понимания картины опубликую все. 

1.) Проверка наличия директорий и их создание:
---
- hosts: all
  gather_facts: no
  become_method: sudo
  become: true
  become_user: ansible
  tasks:
    - name: "Directories exists?"
      file:
        state: directory
        mode: 0755
        group: ansible
        owner: ansible
        path: "{{ item }}"
      loop:
        - /home/ansible/ssl
        - /home/ansible/ssl/example_1.com
        - /home/ansible/ssl/example_2.com
        - /home/ansible/ssl/example_3.com

2.) Копирование сертификатов, создание ссылки, и reload конфигурации nginx:
---
- hosts: all
  gather_facts: no
  become_method: sudo
  become: true
  become_user: ansible
  tasks:
    - name: make sure SSL certificates are existing and secured
      copy:
        src: "/tmp/semaphore/repository_1_1/deploy_ssl/v_1/ssl/{{ item }}/domain.crt"
        dest: "/home/ansible/ssl/{{ item }}/domain.crt"
        owner: ansible
        group: ansible
        mode: "0644"
      loop:
        - example_1.com
        - example_2.com
        - example_3.com

    - name: make sure SSL private keys are installed
      copy:
        src: "/tmp/semaphore/repository_1_1/deploy_ssl/v_1/ssl/{{ item }}/domain.key"
        dest: "/home/ansible/ssl/{{ item }}/domain.key"
        owner: ansible
        group: ansible
        mode: "0644"
      no_log: true
      loop:
        - example_1.com
        - example_2.com
        - example_3.com

    - name: Create symbolic link
      file:
        src: "/home/ansible/ssl"
        dest: "/etc/nginx/ssl"
        state: link
        
    - name: reload nginx
      service:
        name: nginx.service
        state: reloaded

3.) Собственно запускаемый playbook:
---
- name: check and create directory
  import_playbook: check_directory_and_create.yml
    
- name: copy ssl to host
  import_playbook: copy_ssl_to_host.yml

**Comment by fiftin** [comment]
Я уже понял что проблема не в директории. Обновил свой ответ, он перед вашим. Сравните версии python и ansible.  https://www.tecmint.com/fix-shared-connection-to-x-x-xx-closed-ansible-error/

**Comment by DevOpsSim** [comment]
В docker semaphore:

~ $ ansible --version
ansible [core 2.14.5]
  config file = None
  configured module search path = ['/home/semaphore/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.11/site-packages/ansible
  ansible collection location = /home/semaphore/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.11.8 (main, Feb 19 2024, 22:58:08) [GCC 12.2.1 20220924] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
~ $ python --version
Python 3.11.8

На удаленном хосте:
 which python3
/usr/bin/python3

 python3 --version
Python 3.9.2

Т.е версии python в докере semaphore и на удаленном хосте должны совпадать? Правильно ли я Вас понял?

**Comment by fiftin** [comment]
Судя по тому, что я нашел в гугле по ошибке, которую вы указали выше, Ansible-у не удается найти Python на удаленном сервере. Я не знаю почему. Вот такой вариант предлагают в статье, ссылку на которую я указал выше:
```ini
[prod_servers]
192.168.10.20			ansible_python_interpreter=/usr/bin/python3.6
```

Явным образом указать путь к Python на удаленном сервере.

Но это никак не относится в теме issue :) Это другая проблема, связанная с окружением, а не с Semaphore.

**Comment by DevOpsSim** [comment]
@fiftin Да, Вы действительно правы! Добавил в окружение параметры {"interpreter_python" : "python3"}. Предыдущая ошибка ушла. Но Ошибка доступа:  fatal: [test-debian]: FAILED! => {"changed": false, "msg": "Error while linking: [Errno 13] Permission denied: b'/home/ansible/ssl' -> b'/etc/nginx/ssl'", "path": "/etc/nginx/ssl"}   По-прежнему существует.

**Comment by fiftin** [comment]
Потому что вы делаете так `become_user: ansible`: т е `sudo -iu ansible`, хотя вы и так подключаетесь как `ansible`. Этот параметр не нужен, если вы хотите выполнять команду из под `root`. Ну или я чего-то не понял :)

**Comment by fiftin** [comment]
@DevOpsSim я нашел баг, который, возможно с проблемой, которую вы описали.

Т е вы подключаетесь к серверу не как `ansible`, а потом делаете `become` как `ansible` и у него есть root-права?

@DevOpsSim вы можете присоединиться к Телеграм-группе https://t.me/semaphoreui, возможно там оперативнее будет.

**Comment by fiftin** [comment]
@DevOpsSim я отправил на сборку фикс. Скоро будет

**Comment by fiftin** [comment]
@DevOpsSim Fix ready [v2.9.69-beta](https://github.com/semaphoreui/semaphore/tree/v2.9.69-beta)

**Comment by ipa-rar** [comment]
@fiftin   I tried v.9.69-beta and tried to Update and upgrade apt packages with this script

```
---
- hosts: all
  become: true
  become_user: root
  become_method: sudo
  connection: local
  tasks:
    - name: Update and upgrade apt packages
      ansible.builtin.apt:
        update_cache: yes
        upgrade: yes
```

`fatal: [192.168.50.156]: FAILED! => {"msg": "privilege output closed while waiting for password prompt:\n/bin/sh: sudo: not found\n"}`

This happens to all the playbooks that I have with `become=true`
